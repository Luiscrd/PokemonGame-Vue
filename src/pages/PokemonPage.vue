<template>
    <div class="titulo">
        <h1>¿Quién es este Pokémon?</h1>
    </div>
    <div v-if="!pokemon" class="container mt-5 text-center">
        <h1>Espere por favor...</h1>
    </div>
    <div v-else class="container mt-5 text-center">
        <PokemonPicture :pokemonId="pokemon.id" :showPokemon="showPokemon"/>
        <PokemonOptions :pokemons="pokemonArr" @selection="checkAnswer"/>
        <div v-if="showAnswer" class="mb-5">
            <h2 class="mb-5">{{ message }}</h2>
            <button type="button" class="btn btn-primary" @click="newGame">Nuevo juego</button>
        </div>
    </div>
  
</template>

<script>
import PokemonPicture from '@/components/pokemon/PokemonPicture.vue'
import PokemonOptions from '@/components/pokemon/PokemonOptions.vue'
import getPokemonOptions from '@/helpers/getPokemonsOptions'


export default {
  components: { PokemonPicture },
    name: 'PokemonPage',
    components: {
        PokemonPicture,
        PokemonOptions
    },
    data() {
        return {
            pokemonArr: [],
            pokemon: null,
            showPokemon: false,
            showAnswer: false,
            message: ''
        }
    },
    methods: {
        async mixPokemonArray() {
            this.pokemonArr = await getPokemonOptions()
            const rndInt = Math.floor( Math.random() * 4 )
            this.pokemon = this.pokemonArr[ rndInt ]
        },
        checkAnswer( selectedId ) {
            this.showPokemon = true
            this.showAnswer = true

            if( selectedId === this.pokemon.id ) {
                this.message = `Correcto, ${ this.pokemon.name }!!`
            } else {
                this.message = `Upss, era ${ this.pokemon.name }!!`
            }
        },
        newGame() {
            this.showAnswer = false
            this.showPokemon = false
            this.pokemonArr = []
            this.pokemon = null
            this.mixPokemonArray()
        }
    },
    mounted() {
        this.mixPokemonArray()
    }
}
</script>

<style scoped>
.titulo{
    background-color: #263238;
    color: white;
    padding: 40px 0;
    text-align: center;
}

.btn:hover {
    color: white;
    background-color: #13191c;
    border-color: #13191c;
}
.btn-primary {
    --bs-btn-color: #fff;
    --bs-btn-bg: #263238;
    --bs-btn-border-color: #263238;
    --bs-btn-hover-color: #fff;
    --bs-btn-hover-bg: #263238;
    --bs-btn-hover-border-color: #263238;
    --bs-btn-focus-shadow-rgb: 49,132,253;
    --bs-btn-active-color: #fff;
    --bs-btn-active-bg: #263238;
    --bs-btn-active-border-color: #263238;
    --bs-btn-active-shadow: inset 0 3px 5px rgba(0, 0, 0, 0.125);
    --bs-btn-disabled-color: #fff;
    --bs-btn-disabled-bg: #263238;
    --bs-btn-disabled-border-color: #263238;
}

h2{
    color: #EF5350;
}
</style>
