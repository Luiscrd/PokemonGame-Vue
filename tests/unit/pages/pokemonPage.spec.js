import { mount, shallowMount } from '@vue/test-utils'
import PokemonPage from '@/pages/PokemonPage.vue'
import { mockPokemons } from '../mocks/pokemons.mock'

describe('PokemonOptions Component', () => {

    let wrapper
    let mixPokemonArraySpy

    beforeEach(() => {

        mixPokemonArraySpy = jest.spyOn(PokemonPage.methods, 'mixPokemonArray')

        wrapper = mount(PokemonPage, {
            data() {
                return {
                    pokemonArr: mockPokemons,
                    pokemon: mockPokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

    });


    test('Debe de llamar mixPokemonArray al montar', () => {


        expect(mixPokemonArraySpy).toHaveBeenCalled()

    })

    test('Debe de hacer mach con el snapshot cuando se cargan los PokÃ©mons', () => {

        expect(wrapper.html()).toMatchSnapshot()

    })

    test('Debe de mostrar los componentes de PokemonPicture y PokemonOptions', () => {

        wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: mockPokemons,
                    pokemon: mockPokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        expect(wrapper.find('pokemon-picture-stub').exists()).toBeTruthy()

        expect(wrapper.find('pokemon-options-stub').exists()).toBeTruthy()

        expect(wrapper.find('pokemon-picture-stub').attributes('pokemonid')).toBe('1')

        expect(wrapper.find('pokemon-options-stub').attributes('pokemons')).toBeTruthy()

    })

    test('Pruebas con el checkAnswer', async () => {

        wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: mockPokemons,
                    pokemon: mockPokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        await wrapper.vm.checkAnswer(1)

        expect(wrapper.find('h2').exists()).toBeTruthy()

        expect(wrapper.vm.showPokemon).toBeTruthy()

        expect(wrapper.find('h2').text()).toBe('Correcto, bulbasaur!!')

        await wrapper.vm.checkAnswer(2)

        expect(wrapper.find('h2').text()).toBe('Upss, era bulbasaur!!')


    })


})