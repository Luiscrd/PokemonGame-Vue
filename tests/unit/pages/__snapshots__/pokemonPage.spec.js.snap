// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonOptions Component Debe de hacer mach con el snapshot cuando se cargan los Pokémons 1`] = `
<div class="titulo">
  <h1>¿Quién es este Pokémon?</h1>
</div>
<div class="container mt-5 text-center">
  <div class="pokemon-container mt-3"><img src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/1.svg" class="hidden-pokemon" alt="pokemon">
    <!--v-if-->
  </div>
  <div class="options-container mt-5 mb-4">
    <ul>
      <li><button type="button" class="btn btn-primary">bulbasaur</button></li>
      <li><button type="button" class="btn btn-primary">ivysaur</button></li>
      <li><button type="button" class="btn btn-primary">venusaur</button></li>
      <li><button type="button" class="btn btn-primary">charmander</button></li>
    </ul>
  </div>
  <!--v-if-->
</div>
`;
